datalad/local/foreach.py

```diff
@@ -0,0 +1,275 @@
+# emacs: -*- mode: python; py-indent-offset: 4; tab-width: 4; indent-tabs-mode: nil -*-
+# ex: set sts=4 ts=4 sw=4 noet:
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+#
+#   See COPYING file distributed along with the datalad package for the
+#   copyright and license terms.
+#
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+"""Plumbing command for running a command on each (sub)dataset"""
+
+__docformat__ = 'restructuredtext'
+
+
+import logging
+
+from argparse import REMAINDER
+from itertools import chain
+from tempfile import mkdtemp
+
+from datalad.cmd import NoCapture, StdOutErrCapture
+from datalad.core.local.run import normalize_command
+
+from datalad.distribution.dataset import (
+    Dataset,
+    EnsureDataset,
+    datasetmethod,
+    require_dataset,
+)
+from datalad.dochelpers import exc_str
+from datalad.interface.base import (
+    Interface,
+    build_doc,
+)
+from datalad.interface.common_opts import (
+    jobs_opt,
+    recursion_flag,
+    recursion_limit,
+)
+from datalad.interface.results import get_status_dict
+from datalad.interface.utils import eval_results
+from datalad.support.constraints import (
+    EnsureBool,
+    EnsureNone,
+)
+from datalad.support.parallel import (
+    ProducerConsumer,
+    ProducerConsumerProgressLog,
+    no_parentds_in_futures,
+    no_subds_in_futures,
+)
+from datalad.support.param import Parameter
+from datalad.utils import (
+    SequenceFormatter,
+    getpwd,
+    shortened_repr,
+)
+
+lgr = logging.getLogger('datalad.local.foreach')
+
+
+@build_doc
+class ForEach(Interface):
+    r"""Run a command on the dataset and/or each of its sub-datasets.
+
+    Each dataset's repo runner is reused to execute the commands, so they
+    are executed in the top directory of a corresponding dataset.
+    In contrast, Python commands (expressions) are evaluated without changing directory.
+    TODO: unify via explicit --chdir? but not sure if worth it
```

Comment:
not sure what "normal rule" here is but yeah, I think I am of the same opinion now -- I will make them consistently `chpwd` . Later we might like to add an option `--chpwd=True|False` to facilitate possible cases where it is desired to operate from current directory.

Reply:
as of now -- no difference, and in both (external vs python) execution should happen while in that directory...
BUT it stroke me while writing that it would not work AFAIK with `-J` since we are parallelizing via threads, and `chpwd` would change PWD (AFAIK) of the entire process.  Added to TODOs and an example of "fail" under "can't chpwd for Python eval/exec when parallelizing using threads" in the description.
So would need to either revert to prior behavior (easy, and for python since `ds.` operations do not require chdir'ing I think it is generally not that bad), or figure out may be a way to run them in a subprocess, but then see yet another [known async gotcha](https://github.com/datalad/datalad/pull/5369/files#diff-4aa1a9688db72f7a475890e0ee6a343ea98af70ccc3bfa66f17da82c819dfa8dR188) so might be not that trivial ATM

Reply:
I mentally boiled down to
- have `--chpwd=True|False` (so consistent between Python and CLI)
- crash with `NotImplementedError` if python and jobs discovered to be > 1: we have other non-thread safe calls (as e.g. `getcwd()` stumbled upon in #5606), and supporting parallel execution through threads could lead to surprises (and I just now realized that the whole threads-parallelism might be in jeopardy! need to check :-/).  so we need to approach it only with careful analysis etc.

Reply:
decided against NotImplementedError to not cause pain and suffering whenever default `jobs` is set globally -- went for just issuing a warning and falling back to a threadless run.  Also the same in the case of `capture` mode for outputs -- we can't use `swallow_outputs` across threads really since they all use the same std streams.